File: LintCode_Python/topKFrequentWords.py
# Given a list of words and an integer k, return the top k frequent words in the
# list.

# Have you met this question in a real interview? Yes
#  Notice

# You should order the words by the frequency of them in the return list, the most
#  frequent one comes first. If two words has the same frequency, the one with
#  lower alphabetical order come first.

# Example
# Given

# [
#     "yes", "lint", "code",
#     "yes", "code", "baby",
#     "you", "baby", "chrome",
#     "safari", "lint", "code",
#     "body", "lint", "code"
# ]
# for k = 3, return ["code", "lint", "baby"].

# for k = 4, return ["code", "lint", "baby", "yes"],

# Challenge
# Do it in O(nlogk) time and O(n) extra space.

# Extra points if you can do it in O(n) time with O(k) extra space.

# from collections import Counter

def cmp(a, b):
    if a[0] > b[0] or (a[0] == b[0] and a[1] < b[1]):
        return -1
    elif a[0] == b[0] and a[1] == b[1]:
        return 0
    else:
        return 1

def topKFrequentWords(words, k):

    # Method 1:
    # d = dict()
    # for word in words:
    #     if word in d:
    #         d[word] += 1
    #     else:
    #         d[word] = 1
    # # print d
    # result = []
    # count = 0
    # while count < k:
    #     freq = 0
    #     freqWord = "test"
    #     for word in d:
    #         if d[word] > freq or (d[word] == freq and word < freqWord):
    #             freq = d[word]
    #             freqWord = word
    #     del d[freqWord]
    #     result.append(freqWord)
    #     count += 1
    # return result


    # Method 2:
    d = dict()
    for word in words:
        if word in d:
            d[word] += 1
        else:
            d[word] = 1
    p = []
    for key, value in d.items():
        p.append([value, key])
    p.sort(cmp=cmp)
    result = []
    for i in range(k):
        result.append(p[i][1])
    return result



# print topKFrequentWords(["ba","fe","bd","bf","fe","ae","ae","ed","ef","ab",
    # "cd","ac","cf","af","ed","ef","fb","ba","dc","ca","cb","ca","bc","bf","ae",
    # "ec","fa","ac","bd","af","fa","dc","cb","ed","db","fa","cb","fe","ab","bd",
    # "eb","af","ad","cd","bf","bc","cd","fd","de","fc","ef","bc","ca","fe","ac",
    # "ac","cb","eb","ca","fa","bf","ed","cb","ef","be","de","da","bc","ad","cf",
    # "ef","fd","ce","be","df","bf","fd","ef","ab","ef"], 3)
# print Counter(l)
print topKFrequentWords([
    "yes", "lint", "code",
    "yes", "code", "baby",
    "you", "baby", "chrome",
    "safari", "lint", "code",
    "body", "lint", "code"
], 4)

File: LeetCode_Java/src/TopKFrequentWords.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

/**
 * Given a non-empty list of words, return the k most frequent elements.
 * Your answer should be sorted by frequency from highest to lowest. If two words have the same 
 * frequency, then the word with the lower alphabetical order comes first.
 * Example 1:
 * Input: ["i", "love", "leetcode", "i", "love", "coding"], k = 2
 * Output: ["i", "love"]
 * Explanation: "i" and "love" are the two most frequent words. 
 * Note that "i" comes before "love" due to a lower alphabetical order.
 * Example 2:
 * Input: ["the", "day", "is", "sunny", "the", "the", "the", "sunny", "is", "is"], k = 4
 * Output: ["the", "is", "sunny", "day"]
 * Explanation: "the", "is", "sunny" and "day" are the four most frequent words,
 *     with the number of occurrence being 4, 3, 2 and 1 respectively.
 * Note:
 * 1. You may assume k is always valid, 1 ≤ k ≤ number of unique elements.
 * 2. Input words contain only lowercase letters.
 * Follow up:
 * 1. Try to solve it in O(n log k) time and O(n) extra space.
 * @author wendi
 *
 */
public class TopKFrequentWords {
	
	/**
	 * Approach1: bucket sort
	 * @param String[] words, int k
	 * @return List<String>
	 * Time: O(n)
	 * Space: O(n)
	 */
	public List<String> topKFrequentWordsI(String[] words, int k) {
        List<String> res = new ArrayList<>();
        if (words == null || words.length == 0 || k == 0) return res;
        Map<String, Integer> map = new HashMap<>();
        int max = 0;
        for (String word: words) {
            map.put(word, map.getOrDefault(word, 0) + 1);
            max = Math.max(map.get(word), max);
        }
        List<String>[] bucket = new List[max + 1];
        for (String word: map.keySet()) {
            int cnt = map.get(word);
            if (bucket[cnt] == null) bucket[cnt] = new ArrayList<>();
            bucket[cnt].add(word);
        }
        for (int i = max; i >= 0 && k > 0; i--) {
            if (bucket[i] == null) continue;
            Collections.sort(bucket[i]);
            for (int j = 0; j < bucket[i].size() && k > 0; j++) {
                res.add(bucket[i].get(j));
                k--;
            }
        }
        return res;		
	}
	
	
	/**
	 * Approach1: minHeap
	 * Count the frequency of each word, then add it to min heap that stores the top frequent k 
	 * candidates. At the end, we pop off the heap up to k times and reverse the result.
	 * @param String[] words, int k
	 * @return List<String>
	 * Time: O(nlog(k))
	 * Space: O(n)
	 */
	public List<String> topKFrequentWords(String[] words, int k) {
		List<String> result = new ArrayList<>();
		if (words == null || words.length == 0) return result;
		Map<String, Integer> freq = new HashMap<>();
		for (String word: words) {
			if (!freq.containsKey(word)) freq.put(word, 0);
			freq.put(word, freq.get(word) + 1);
		}
		PriorityQueue<Map.Entry<String, Integer>> minHeap = new PriorityQueue<>
		(k, new Comparator<Map.Entry<String, Integer>>() {
			@Override
			public int compare(Map.Entry<String, Integer> e1, Map.Entry<String, Integer> e2) {
				if (e1.getValue() != e2.getValue()) return e1.getValue() - e2.getValue();
				return e2.getKey().compareTo(e1.getKey());
			}
		});
		for (Map.Entry<String, Integer> e: freq.entrySet()) {
			minHeap.offer(e);
			if (minHeap.size() > k) minHeap.poll();
		}
		while (k-- > 0) {
			result.add(minHeap.poll().getKey());
		}
		Collections.reverse(result);
		return result;
	}

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		TopKFrequentWords result = new TopKFrequentWords();
		System.out.println(result.topKFrequentWords(new String[] {"i", "love", "leetcode", "i", "love", "coding"}, 2));
		System.out.println(result.topKFrequentWordsI(new String[] {"i", "love", "leetcode", "i", "love", "coding"}, 2));

	}

}
